alf/summary/render.py

```diff
@@ -0,0 +1,658 @@
+
+    return im, cbar
+
+
+def _annotate_heatmap(im,
+                      valfmt="%.2f",
+                      textcolors=("black", "white"),
+                      threshold=None,
+                      **textkw):
+    """A function to annotate a heatmap.
+
+    (Code from `matplotlib documentation <https://matplotlib.org/stable/gallery/images_contours_and_fields/image_annotated_heatmap.html>`_)
+
+    Args:
+        im (matplotlib.image.AxesImage): The image to be labeled.
+        valfmt (str): The format of the annotations inside the heatmap. This
+            should either use the string format method, e.g. "%.2f", or be
+            a ``matplotlib.ticker.Formatter``.
+        textcolors (tuple[str]): A pair of colors. The first is used for values
+            below a threshold, the second for those above.
+        threshold (float): Value in data units according to which the colors
+            from textcolors are applied. If None (the default) uses the middle
+            of the colormap as separation.
+        **textkw: All other arguments are forwarded to each call to ``text()``
+            used to create the text labels.
+    """
+    data = im.get_array()
+
+    # Normalize the threshold to the images color range.
+    if threshold is not None:
+        threshold = im.norm(threshold)
+    else:
+        threshold = im.norm(data.max()) / 2.
+
+    # Set default alignment to center, but allow it to be
+    # overwritten by textkw.
+    kw = dict(horizontalalignment="center", verticalalignment="center")
+    kw.update(textkw)
+
+    # Loop over the data and create a `Text` for each "pixel".
+    # Change the text's color depending on the data.
+    for i in range(data.shape[0]):
+        for j in range(data.shape[1]):
+            kw.update(color=textcolors[int(im.norm(data[i, j]) > threshold)])
+            im.axes.text(j, i, valfmt % data[i, j], **kw)
+
+
+@_rendering_wrapper
+def render_heatmap(name,
+                   data,
+                   val_label="",
+                   row_labels=None,
+                   col_labels=None,
+                   cbar_kw={},
+                   annotate_format="%.2f",
+                   font_size=7,
+                   img_height=256,
+                   img_width=256,
+                   dpi=300,
+                   figsize=(2, 2),
+                   **kwargs):
+    """Render a 2D tensor as a heatmap.
+
+    Args:
+        name (str): rendering identifier
+        data (Tensor|np.ndarray): a tensor/np.array of shape ``[H,W]``
+        val_label (str): The label for the rendered values.
+        row_labels (list[str]): A list of length ``H`` with the labels for
+            the rows.
+        col_labels (list[str]): A list of length ``W`` with the labels for
+            the columns.
+        cbar_kw (dict): A dictionary with arguments to ``matplotlib.Figure.colorbar``.
+        annotate_format (str): The format of the annotations on the heatmap.
```

Comment:
not clear what annotation does.

Reply:
annotation is for overlaying a number on each heatmap slot.